x/programs/rust/wasmlanche-sdk/src/memory.rs

```diff
@@ -7,12 +7,28 @@
 //! the program. These methods are unsafe as should be used
 //! with caution.
 
-use std::{alloc::Layout, cell::RefCell, collections::HashMap, mem::ManuallyDrop, ops::Deref};
+extern crate alloc;
 
-thread_local! {
-    /// Map of pointer to the length of its content on the heap
-    static ALLOCATIONS: RefCell<HashMap<*const u8, usize>> = RefCell::new(HashMap::new());
-}
+use alloc::{
+    alloc::{alloc as allocate, dealloc as deallocate, handle_alloc_error, Layout},
+    vec::Vec,
+};
+use core::{
+    cell::{LazyCell, RefCell},
+    mem::ManuallyDrop,
+    ops::Deref,
+    slice,
+};
+use hashbrown::HashMap;
+
+/// Map of pointer to the length of its content on the heap
+
+// TODO:
+// `RefCell` shouldn't be necessary here since we need unsafe access anyway.
+// However, it does guarantee no UB on a single-thread.
+
+static mut ALLOCATIONS: LazyCell<RefCell<HashMap<*const u8, usize>>> =
```

Comment:
what does `LazyCell` do here?

Reply:
It instantiates on first access because `HashMap::new` has to allocate (not a const function).